web/src/main/java/com/epam/rft/atsy/web/controllers/rest/StateHistoryController.java

```diff
@@ -0,0 +1,95 @@
+package com.epam.rft.atsy.web.controllers.rest;
+
+import com.epam.rft.atsy.service.StatesHistoryService;
+import com.epam.rft.atsy.service.domain.states.StateDTO;
+import com.epam.rft.atsy.service.domain.states.StateHistoryDTO;
+import com.epam.rft.atsy.web.StateHistoryViewRepresentation;
+import com.epam.rft.atsy.web.exceptionhandling.RestResponse;
+import org.springframework.context.MessageSource;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Collections;
+import java.util.Locale;
+import javax.annotation.Resource;
+import javax.validation.Valid;
+
+@RestController
+@RequestMapping(value = "/secure/application_state")
+public class StateHistoryController {
+
+  private static final String DATE_FORMAT_CONSTANT = "yyyy-MM-dd HH:mm:ss";
+
+  private static final String COMMON_INVALID_INPUT_MESSAGE_KEY = "common.invalid.input";
+
+  @Resource
+  private StatesHistoryService statesHistoryService;
+
+  @Resource
+  private MessageSource messageSource;
+
+  @RequestMapping(method = RequestMethod.POST)
+  public ResponseEntity saveOrUpdate(@RequestParam Long applicationId,
+                                     @Valid @RequestBody StateHistoryViewRepresentation stateHistoryViewRepresentation,
+                                     BindingResult bindingResult, Locale locale) {
+
+    SimpleDateFormat simpleDateFormat = new SimpleDateFormat(DATE_FORMAT_CONSTANT);
+
+    StateHistoryDTO stateHistoryDTO = null;
+
+    if (bindingResult.hasErrors()) {
+      String
```

Comment:
It seems that this is a generic algorithm which can be used at other controllers as well.
You can extract this functionality into a separate class and inject it as a dependency.